engine/grid.py

```diff
@@ -331,36 +337,78 @@ def find_border_nodes(self):
             is_vertical = False
+
         border_list = []
         leftmost_node = None
-        leftmost_node_pos = None
-
+        bottom_left_pos = None
+        top_left_pos = None
+        bottom_right_pos = None
+        top_right_pos = None
+        # go through each row and column
         for row in range(self.num_rows):
             for col in range(self.num_cols):
                 node = self.nodes[row][col]
+                # check if this is an active node and on the border
                 if node.is_active and self.is_on_border(
                     (row, col), self.num_rows, self.num_cols
                 ):
-                    # check if this is an active node and on the border
+                    # add node to list of border nodes
                     self.nodes[row][col].is_border = True
                     border_list.append((node, row, col))
+                    # if the bottom_left_pos is uninitialized or
+                    # if we are traversing vertically, keep shifting the point to the left
+                    # if we are traversing horizontally, keep shifting the point downwards
+                    # This logic is necessary for cases like the circle, where the initial
+                    # position is different depending on if we are traversing vertically or horizonally
                     if (
-                        leftmost_node_pos is None
-                        or (not is_vertical and col < leftmost_node_pos[1])
-                        or (is_vertical and row < leftmost_node_pos[0])
+                        bottom_left_pos is None
+                        or (not is_vertical and col < bottom_left_pos[1])
+                        or (is_vertical and row < bottom_left_pos[0])
                     ):
                         leftmost_node = node
-                        leftmost_node_pos = (row, col)
+                        bottom_left_pos = (row, col)
+
+        bottom_right_pos = bottom_left_pos
+        l_col = bottom_left_pos[1]
+        for row in range(self.num_rows):
+            node = self.nodes[row][l_col]
+            if row > bottom_right_pos[0]:
+                if node.is_active and self.is_on_border(
+                    (row, l_col), self.num_rows, self.num_cols
+                ):
+                    bottom_right_pos = (row, l_col)
+
+        top_left_pos = bottom_left_pos
+        top_right_pos = bottom_right_pos
+        l_row = top_left_pos[0]
+        r_row = top_right_pos[0]
+        for col in range(self.num_cols):
+            node = self.nodes[l_row][col]
+            if col > top_left_pos[1]:
+                if node.is_active and self.is_on_border(
+                    (l_row, col), self.num_rows, self.num_cols
+                ):
+                    top_left_pos = (l_row, col)
+                if node.is_active and self.is_on_border(
+                    (r_row, col), self.num_rows, self.num_cols
+                ):
+                    top_right_pos = (r_row, col)
```

Comment:
I dont think this works if the shape is wide on the bottom and narrow at the top (like a triangle)